silkworm/node/stagedsync/stages/stage_triggers.cpp

```diff
@@ -40,13 +40,13 @@ Stage::Result TriggersStage::forward(db::RWTxn& tx) {
 }
 
 Task<void> TriggersStage::schedule(std::function<void(db::RWTxn&)> callback) {
-    auto task_caller = [this, c = std::move(callback)]() -> Task<void> {
+    auto task_caller = [this, trigger = std::move(callback)]() -> Task<void> {
         db::RWTxn* tx = this->current_tx_;
         SILKWORM_ASSERT(tx);
-        c(*tx);
+        trigger(*tx);
         co_return;
     };
-    return concurrency::spawn_task(ioc_, task_caller());
+    co_return co_await concurrency::spawn_task(ioc_, task_caller());
```

Comment:
This is an interesting side effect. I wonder why this change makes any difference?

The call to schedule is already awaited [here on the call site](https://github.com/erigontech/silkworm/blob/master/silkworm/db/freezer.cpp#L164)

An additional await here delays the destruction of some locals, right? Which ones? `task_caller` is not relevant after calling it. `trigger` is moved into the task, so not destroyed unless the spawned task is destroyed. The spawned task is moved inside co_spawn. I hope that co_spawn is not destroying the passed task prematurely, but maybe it does? Is it possible to verify with a unit test?

If it really does, maybe we modify spawn_task to await for co_spawn inside spawn_task so that this mistake is not possible in other places (we use spawn_task without awaiting in a bunch of places) and in the future code?

I'd like to fill the gap in my understanding here. I know it is not the high priority, so let's maybe create an issue about researching and generalizing this and make it an exercise for the future reader?

Reply:
This could potentially be a reason of some sentry discovery crashes, because SerialNodeDb follows this pattern.

Reply:
let's try to move the callback from the capture list to the lambda parameter list and see if it works without co_await (see links below)

Reply:
Yep, we must move both `this` and the callback because all the capture list is affected (indeed, the issue was happening on `this->current_tx_`).
Moreover, it seems to be reproducible systematically in unit tests, so I'm going to add one test and fix by moving from capture list to the lambda parameters.